Tests/OptimizelyTests-DataModel/OptimizelyJSONTests.swift

```diff
@@ -0,0 +1,489 @@
+    }
+    
+    private struct Field4: Decodable, Equatable {
+        
+        static func == (lhs: OptimizelyJSONTests.Field4, rhs: OptimizelyJSONTests.Field4) -> Bool {
+            if (lhs.innerField1 != rhs.innerField1 || lhs.innerField2.count != rhs.innerField2.count) {
+                return false
+            }
+            for (index,value) in lhs.innerField2.enumerated() {
+                if rhs.innerField2[index] != value {
+                    return false
+                }
+            }
+            return true
+        }
+        
+        var innerField1: Int = 0
+        var innerField2: [InnerField2] = []
+        
+        enum CodingKeys: String, CodingKey {
+            case innerField1 = "inner_field1"
+            case innerField2 = "inner_field2"
+        }
+    }
+    
+    private enum InnerField2: Decodable, Equatable  {
+        case bool(Bool)
+        case double(Double)
+        case string(String)
+        
+        init(from decoder: Decoder) throws {
+            let container = try decoder.singleValueContainer()
+            if let x = try? container.decode(Bool.self) {
+                self = .bool(x)
+                return
+            }
+            if let x = try? container.decode(Double.self) {
+                self = .double(x)
+                return
+            }
+            if let x = try? container.decode(String.self) {
+                self = .string(x)
+                return
+            }
+            throw DecodingError.typeMismatch(InnerField2.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Wrong type for InnerField2"))
+        }
+    }
+    
+    private struct EmptySchema: Decodable {
+    }
+    
+    private struct IncorrectSchema: Decodable {
+        var incorrectField: Int = 0
+    }
+    
+    private var payload = ""
+    private var data = [String: Any]()
+    private var innerField2List = [InnerField2]()
+    private var optimizelyJSON: OptimizelyJSON!
+    
+    override func setUp() {
+        self.innerField2List = [
+            InnerField2.string("1"),
+            InnerField2.string("2"),
+            InnerField2.double(3.01),
+            InnerField2.double(4.23),
+            InnerField2.bool(true)
+        ]
+        self.payload = """
+        {
+        "field1": 1,
+        "field2": 2.5,
+        "field3": "three",
+        "field4": {"inner_field1":3,"inner_field2":["1","2",3.01,4.23,true]},
+        "field5": true,
+        }
+        """
+        self.data = [
+            "field1": 1,
+            "field2": 2.5,
+            "field3": "three",
+            "field4": [
+                "inner_field1": 3.0,
+                "inner_field2": ["1","2",3.01,4.23,true]
+            ],
+            "field5": true,
+        ]
+        try! self.optimizelyJSON = OptimizelyJSON(payload: self.payload)
+    }
+}
+
+// MARK: - Tests
+
+extension OptimizelyJSONTests {
+    
+    func testConstructorsHappyPath() {
```

Comment:
Please break down into multiple cases. it's a long test case verifying multiple things.